File: typescript/tests/payment/application/services/refundHandling.test.ts
import { beforeEach, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Description } from '@/common/domain/primitive/description.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { RefundHandling } from '@/payment/application/services/refundHandling.js';
import { PaymentEvent } from '@/payment/domain/events/paymentEvent.js';
import {
  refundRequested,
  type RefundRequested,
} from '@/payment/domain/events/refundRequested.js';
import type { TransactionRefunded } from '@/payment/domain/events/refundResultEvents.js';
import { Payment } from '@/payment/domain/payment/payment.js';
import { PaymentId } from '@/payment/domain/payment/paymentId.js';
import { PaymentStatus } from '@/payment/domain/payment/paymentStatus.js';
import { providerFromLabel } from '@/payment/domain/payment/provider.js';
import { ProviderReference } from '@/payment/domain/payment/providerReference.js';
import { TransactionId } from '@/payment/domain/payment/transactionId.js';
import { TransactionStatus } from '@/payment/domain/payment/transactionStatus.js';
import { TransactionRefund } from '@/payment/domain/policies/transactionRefund.js';
import type { PaymentProvider } from '@/payment/domain/ports/paymentProvider.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import {
  paymentProviderFailure,
  paymentProviderSuccess,
} from '@/payment/domain/ports/paymentProviderResult.js';
import { SqlitePaymentRepository } from '@/payment/infrastructure/sqlitePaymentRepository.js';
import { CapturingEventPublisher } from '../../../testsupport/events/capturingEventPublisher.js';
import { DatabaseSetup } from '../../../testsupport/databaseSetup.js';

describe('RefundHandling', () => {
  let repository: SqlitePaymentRepository;
  let publisher: CapturingEventPublisher<PaymentEvent>;
  let refundHandling: RefundHandling;

  beforeEach(() => {
    const database = DatabaseSetup.initializeInMemoryDb('payment');
    repository = new SqlitePaymentRepository(database);
    publisher = new CapturingEventPublisher();

    const alwaysSuccess: PaymentProvider = {
      process: () => paymentProviderSuccess(Uuid.generate(), Timestamp.now()),
      refund: () => paymentProviderSuccess(Uuid.generate(), Timestamp.now()),
    };
    const alwaysFailure: PaymentProvider = {
      process: () => paymentProviderSuccess(Uuid.generate(), Timestamp.now()),
      refund: () => paymentProviderFailure(new Description('refund declined')),
    };

    const providers: Record<string, PaymentProvider> = {
      PayPal: alwaysSuccess,
      Klarna: alwaysFailure,
      GiftCard: alwaysSuccess,
    };

    refundHandling = new RefundHandling(repository, providers, new TransactionRefund(), publisher);
  });

  describe('construction', () => {
    it('rejects null repository', () => {
      expect(() => new RefundHandling(null as unknown as SqlitePaymentRepository, {}, new TransactionRefund(), publisher)).toThrow();
    });

    it('rejects null providers', () => {
      expect(() => new RefundHandling(repository, null as unknown as Record<string, PaymentProvider>, new TransactionRefund(), publisher)).toThrow();
    });

    it('rejects null policy', () => {
      expect(() => new RefundHandling(repository, {}, null as unknown as TransactionRefund, publisher)).toThrow();
    });

    it('rejects null publisher', () => {
      expect(() => new RefundHandling(repository, {}, new TransactionRefund(), null as unknown as CapturingEventPublisher<PaymentEvent>)).toThrow();
    });
  });

  describe('execution', () => {
    it('rejects null event', () => {
      expect(() => refundHandling.on(null as unknown as RefundRequested)).toThrow();
    });

    it('should refund when provider succeeds', () => {
      const paymentId = seedAcceptedPayment('PayPal', 50);

      refundHandling.on(refundRequested(paymentId, new ClientReference(crypto.randomUUID()), new Money(50)));

      const loaded = repository.findById(paymentId);
      expect(loaded).not.toBeNull();
      expect(loaded!.status()).toEqual(PaymentStatus.REFUNDED);
      expect(loaded!.transactions().every((t) => t.status() === TransactionStatus.REFUNDED)).toBe(true);
      expect(publisher.events()).toHaveLength(1);
      expect(publisher.events()[0].kind).toBe('TransactionRefunded');
      const refunded = publisher.events()[0] as TransactionRefunded;
      expect(refunded.amount).toEqual(new Money(50));
    });

    it('should notify failure when provider rejects', () => {
      const paymentId = seedAcceptedPayment('Klarna', 50);

      refundHandling.on(refundRequested(paymentId, new ClientReference(crypto.randomUUID()), new Money(50)));

      const loaded = repository.findById(paymentId);
      expect(loaded).not.toBeNull();
      expect(loaded!.status()).toEqual(PaymentStatus.ACCEPTED);
      expect(publisher.events()).toHaveLength(1);
      expect(publisher.events()[0].kind).toBe('TransactionNotRefunded');
    });

    it('should refund all accepted transactions', () => {
      const paymentId = seedSplitPayment('PayPal', 60, 'GiftCard', 40);

      refundHandling.on(refundRequested(paymentId, new ClientReference(crypto.randomUUID()), new Money(100)));

      const loaded = repository.findById(paymentId);
      expect(loaded).not.toBeNull();
      expect(loaded!.status()).toEqual(PaymentStatus.REFUNDED);
      expect(loaded!.transactions().every((t) => t.status() === TransactionStatus.REFUNDED)).toBe(true);
      expect(publisher.events()).toHaveLength(1);
      expect(publisher.events()[0].kind).toBe('TransactionRefunded');
      const refunded = publisher.events()[0] as TransactionRefunded;
      expect(refunded.amount).toEqual(new Money(100));
    });

    it('should mark successful transactions as refunded even if one fails', () => {
      const paymentId = seedSplitPayment('PayPal', 60, 'Klarna', 40);

      refundHandling.on(refundRequested(paymentId, new ClientReference(crypto.randomUUID()), new Money(100)));

      const loaded = repository.findById(paymentId);
      expect(loaded).not.toBeNull();
      expect(loaded!.status()).toEqual(PaymentStatus.ACCEPTED);
      const refundedCount = loaded!.transactions().filter((t) => t.status() === TransactionStatus.REFUNDED).length;
      expect(refundedCount).toEqual(1);
      expect(publisher.events()).toHaveLength(1);
      expect(publisher.events()[0].kind).toBe('TransactionNotRefunded');
    });

    it('should fail if payment not found', () => {
      const paymentId = generateId((value) => new PaymentId(value));

      expect(() => refundHandling.on(refundRequested(paymentId, new ClientReference(crypto.randomUUID()), new Money(10)))).toThrow();
    });
  });

  function seedAcceptedPayment(provider: string, amount: number): PaymentId {
    const paymentId = generateId((value) => new PaymentId(value));
    const clientReference = crypto.randomUUID();
    const money = new Money(amount);
    const requestedAt = Timestamp.now();
    const payment = Payment.request(paymentId, new ClientReference(clientReference), money, requestedAt);
    const transactionId = generateId((value) => new TransactionId(value));
    payment.startTransaction(
      transactionId,
      providerFromLabel(provider),
      new ProviderReference(Uuid.generate()),
      money,
      new Timestamp(new Date(requestedAt.value.getTime() + 1000))
    );
    payment.acceptTransaction(transactionId, new Timestamp(new Date(requestedAt.value.getTime() + 60000)));
    repository.save(payment);
    return paymentId;
  }

  function seedSplitPayment(firstProvider: string, firstAmount: number, secondProvider: string, secondAmount: number): PaymentId {
    const paymentId = generateId((value) => new PaymentId(value));
    const clientReference = crypto.randomUUID();
    const total = new Money(firstAmount + secondAmount);
    const requestedAt = Timestamp.now();
    const payment = Payment.request(paymentId, new ClientReference(clientReference), total, requestedAt);

    const firstTx = generateId((value) => new TransactionId(value));
    payment.startTransaction(
      firstTx,
      providerFromLabel(firstProvider),
      new ProviderReference(Uuid.generate()),
      new Money(firstAmount),
      new Timestamp(new Date(requestedAt.value.getTime() + 1000))
    );
    payment.acceptTransaction(firstTx, new Timestamp(new Date(requestedAt.value.getTime() + 60000)));

    const secondTx = generateId((value) => new TransactionId(value));
    payment.startTransaction(
      secondTx,
      providerFromLabel(secondProvider),
      new ProviderReference(Uuid.generate()),
      new Money(secondAmount),
      new Timestamp(new Date(requestedAt.value.getTime() + 2000))
    );
    payment.acceptTransaction(secondTx, new Timestamp(new Date(requestedAt.value.getTime() + 120000)));

    repository.save(payment);
    return paymentId;
  }
});

File: java/src/test/java/it/giannibombelli/wsc2026/payment/application/services/RefundHandlingTest.java
package it.giannibombelli.wsc2026.payment.application.services;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentEvent;
import it.giannibombelli.wsc2026.payment.domain.events.RefundRequested;
import it.giannibombelli.wsc2026.payment.domain.events.RefundResultEvents;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentStatus;
import it.giannibombelli.wsc2026.payment.domain.payment.Provider;
import it.giannibombelli.wsc2026.payment.domain.payment.ProviderReference;
import it.giannibombelli.wsc2026.payment.domain.payment.TransactionId;
import it.giannibombelli.wsc2026.payment.domain.payment.TransactionStatus;
import it.giannibombelli.wsc2026.payment.application.policies.TransactionRefund;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentProvider;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentProviderResult;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;
import it.giannibombelli.wsc2026.payment.infrastructure.SqlitePaymentRepository;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import it.giannibombelli.wsc2026.testsupport.events.CapturingEventPublisher;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class RefundHandlingTest {
    private PaymentRepository repository;
    private CapturingEventPublisher<PaymentEvent> publisher;
    private RefundHandling refundHandling;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("payment");
        repository = new SqlitePaymentRepository(dataSource);
        publisher = new CapturingEventPublisher<>();

        PaymentProvider alwaysSuccess = new PaymentProvider() {
            @Override
            public PaymentProviderResult process(UUID paymentId, UUID providerReference, Money amount) {
                return new PaymentProviderResult.Success(UUID.randomUUID(), Timestamp.now());
            }

            @Override
            public PaymentProviderResult refund(UUID paymentId, UUID providerReference, Money amount) {
                return new PaymentProviderResult.Success(UUID.randomUUID(), Timestamp.now());
            }
        };
        PaymentProvider alwaysFailure = new PaymentProvider() {
            @Override
            public PaymentProviderResult process(UUID paymentId, UUID providerReference, Money amount) {
                return new PaymentProviderResult.Success(UUID.randomUUID(), Timestamp.now());
            }

            @Override
            public PaymentProviderResult refund(UUID paymentId, UUID providerReference, Money amount) {
                return new PaymentProviderResult.Failure(new Description("refund declined"));
            }
        };

        Map<String, PaymentProvider> providers = Map.of(
            "PayPal", alwaysSuccess,
            "Klarna", alwaysFailure,
            "GiftCard", alwaysSuccess
        );

        refundHandling = new RefundHandling(repository, providers, new TransactionRefund(), publisher);
    }

    @Nested
    class Construction {
        @Test
        void rejectNullRepository() {
            assertThatThrownBy(() -> new RefundHandling(null, Map.of(), new TransactionRefund(), publisher))
                .isInstanceOf(NullPointerException.class);
        }

        @Test
        void rejectNullProviders() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            assertThatThrownBy(() -> new RefundHandling(repository, null, new TransactionRefund(), publisher))
                .isInstanceOf(NullPointerException.class);
        }

        @Test
        void rejectNullPolicy() {
            assertThatThrownBy(() -> new RefundHandling(repository, Map.of(), null, publisher))
                .isInstanceOf(NullPointerException.class);
        }

        @Test
        void rejectNullPublisher() {
            assertThatThrownBy(() -> new RefundHandling(repository, Map.of(), new TransactionRefund(), null))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class Execution {
        @Test
        void rejectNullEvent() {
            assertThatThrownBy(() -> refundHandling.on(null))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldRefundWhenProviderSucceeds() {
            PaymentId paymentId = seedAcceptedPayment("PayPal", new BigDecimal("50.00"));

            refundHandling.on(new RefundRequested(paymentId, new ClientReference(UUID.randomUUID()), new Money(new BigDecimal("50.00"))));

            Optional<Payment> loaded = repository.findById(paymentId);
            assertThat(loaded).isPresent();
            assertThat(loaded.get().status()).isEqualTo(PaymentStatus.REFUNDED);
            assertThat(loaded.get().transactions()).allMatch(t -> t.status() == TransactionStatus.REFUNDED);
            assertThat(publisher.events()).hasSize(1);
            assertThat(publisher.events().get(0)).isInstanceOf(RefundResultEvents.TransactionRefunded.class);
            RefundResultEvents.TransactionRefunded refunded = (RefundResultEvents.TransactionRefunded) publisher.events().get(0);
            assertThat(refunded.amount()).isEqualTo(new Money(new BigDecimal("50.00")));
        }

        @Test
        void shouldNotifyFailureWhenProviderRejects() {
            PaymentId paymentId = seedAcceptedPayment("Klarna", new BigDecimal("50.00"));

            refundHandling.on(new RefundRequested(paymentId, new ClientReference(UUID.randomUUID()), new Money(new BigDecimal("50.00"))));

            Optional<Payment> loaded = repository.findById(paymentId);
            assertThat(loaded).isPresent();
            assertThat(loaded.get().status()).isEqualTo(PaymentStatus.ACCEPTED);
            assertThat(publisher.events()).hasSize(1);
            assertThat(publisher.events().get(0)).isInstanceOf(RefundResultEvents.TransactionNotRefunded.class);
        }

        @Test
        void shouldRefundAllAcceptedTransactions() {
            PaymentId paymentId = seedSplitPayment(
                "PayPal", new BigDecimal("60.00"),
                "GiftCard", new BigDecimal("40.00")
            );

            refundHandling.on(new RefundRequested(paymentId, new ClientReference(UUID.randomUUID()), new Money(new BigDecimal("100.00"))));

            Optional<Payment> loaded = repository.findById(paymentId);
            assertThat(loaded).isPresent();
            assertThat(loaded.get().status()).isEqualTo(PaymentStatus.REFUNDED);
            assertThat(loaded.get().transactions()).allMatch(t -> t.status() == TransactionStatus.REFUNDED);
            assertThat(publisher.events()).hasSize(1);
            assertThat(publisher.events().get(0)).isInstanceOf(RefundResultEvents.TransactionRefunded.class);
            RefundResultEvents.TransactionRefunded refunded = (RefundResultEvents.TransactionRefunded) publisher.events().get(0);
            assertThat(refunded.amount()).isEqualTo(new Money(new BigDecimal("100.00")));
        }

        @Test
        void shouldMarkSuccessfulTransactionsAsRefundedEvenIfOneFails() {
            PaymentId paymentId = seedSplitPayment(
                "PayPal", new BigDecimal("60.00"),
                "Klarna", new BigDecimal("40.00")
            );

            refundHandling.on(new RefundRequested(paymentId, new ClientReference(UUID.randomUUID()), new Money(new BigDecimal("100.00"))));

            Optional<Payment> loaded = repository.findById(paymentId);
            assertThat(loaded).isPresent();
            Payment payment = loaded.get();
            assertThat(payment.status()).isEqualTo(PaymentStatus.ACCEPTED);
            long refundedCount = payment.transactions().stream().filter(t -> t.status() == TransactionStatus.REFUNDED).count();
            assertThat(refundedCount).isEqualTo(1);
            assertThat(publisher.events()).hasSize(1);
            assertThat(publisher.events().get(0)).isInstanceOf(RefundResultEvents.TransactionNotRefunded.class);
        }

        @Test
        void shouldFailIfPaymentNotFound() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);

            assertThatThrownBy(() -> refundHandling.on(
                new RefundRequested(paymentId, new ClientReference(UUID.randomUUID()), new Money(new BigDecimal("10.00")))))
                .isInstanceOf(IllegalArgumentException.class);
        }

        private PaymentId seedAcceptedPayment(String provider, BigDecimal amount) {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            UUID clientReference = UUID.randomUUID();
            Money money = new Money(amount);
            Instant requestedAt = Instant.now();
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), money, new Timestamp(requestedAt));
            TransactionId transactionId = EntityId.generate(TransactionId::new);
            payment.startTransaction(
                transactionId,
                Provider.fromLabel(provider),
                new ProviderReference(UUID.randomUUID()),
                money,
                new Timestamp(requestedAt.plusSeconds(1))
            );
            payment.acceptTransaction(transactionId, new Timestamp(requestedAt.plusSeconds(60)));
            repository.save(payment);
            return paymentId;
        }

        private PaymentId seedSplitPayment(String firstProvider, BigDecimal firstAmount,
                                           String secondProvider, BigDecimal secondAmount) {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            UUID clientReference = UUID.randomUUID();
            Money total = new Money(firstAmount.add(secondAmount));
            Instant requestedAt = Instant.now();
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), total, new Timestamp(requestedAt));

            TransactionId firstTx = EntityId.generate(TransactionId::new);
            payment.startTransaction(
                firstTx,
                Provider.fromLabel(firstProvider),
                new ProviderReference(UUID.randomUUID()),
                new Money(firstAmount),
                new Timestamp(requestedAt.plusSeconds(1))
            );
            payment.acceptTransaction(firstTx, new Timestamp(requestedAt.plusSeconds(60)));

            TransactionId secondTx = EntityId.generate(TransactionId::new);
            payment.startTransaction(
                secondTx,
                Provider.fromLabel(secondProvider),
                new ProviderReference(UUID.randomUUID()),
                new Money(secondAmount),
                new Timestamp(requestedAt.plusSeconds(2))
            );
            payment.acceptTransaction(secondTx, new Timestamp(requestedAt.plusSeconds(120)));

            repository.save(payment);
            return paymentId;
        }
    }
}
